src/emulate.c

```diff
@@ -957,25 +962,48 @@ static void match_pattern(riscv_t *rv, block_t *block)
         case rv_insn_lw:
             COMBINE_MEM_OPS(1);
             break;
-        case rv_insn_lui:
-            next_ir = ir + 1;
-            if (next_ir->opcode == rv_insn_addi && ir->rd == next_ir->rs1) {
-                /* The destination register of the LUI instruction is the
-                 * same as the source register 1 of the next instruction ADDI.
-                 */
-                ir->opcode = rv_insn_fuse5;
-                ir->rs1 = next_ir->rd;
-                ir->imm2 = next_ir->imm;
-                ir->impl = dispatch_table[ir->opcode];
-                ir->tailcall = next_ir->tailcall;
-            }
-            break;
             /* TODO: mixture of SW and LW */
             /* TODO: reorder insturction to match pattern */
         }
     }
 }
 
+
+typedef struct {
+    bool is_constant[N_RV_REGS];
+    uint32_t const_val[N_RV_REGS];
+} constopt_info_t;
+
+#define CONSTOPT(inst, code)                                           \
+    static void constopt_##inst(UNUSED rv_insn_t *ir,                  \
+                                UNUSED constopt_info_t *constopt_info) \
+    {                                                                  \
+        code;                                                          \
+    }
+
+#include "rv32_constopt.c"
+/* clang-format off */
+static const void *constopt_table[] = {
+    /* RV32 instructions */
+#define _(inst, can_branch, reg_mask) [rv_insn_##inst] = constopt_##inst,
+    RV_INSN_LIST
+#undef _
+};
+#undef CONSTOPT
+
+/* clang-format on */
+typedef void (*constopt_func_t)(rv_insn_t *, constopt_info_t *);
+static void optimize_constant(block_t *block)
+{
+    constopt_info_t constopt_info;
+    memset(&constopt_info, 0, sizeof(constopt_info));
```

Comment:
The C standards guarantee that even if an integer array is located on automatic storage and if there are fewer initializers in a brace-enclosed list then the uninitialized elements must be initialized to 0.

C99 Standard 6.7.8.21
> If there are fewer initializers in a brace-enclosed list than there are elements or members of an aggregate, or fewer characters in a string literal used to initialize an array of known size than there are elements in the array, the remainder of the aggregate shall be initialized implicitly the same as objects that have static storage duration.

You can safely rewrite as following:
```c
    constopt_info_t info = {0};
```